common/src/project.rs

```diff
@@ -3,146 +3,135 @@ use rustrict::{Censor, Type};
 use serde::de::Error as DeError;
 use serde::{Deserialize, Deserializer, Serialize};
 use std::collections::HashSet;
-use std::error::Error;
-use std::fmt::{Display, Formatter};
+use std::fmt::Formatter;
 use std::str::FromStr;
 
-/// Project names should conform to valid Host segments (or labels)
+/// Project names must conform to valid Host segments (or labels)
 /// as per [IETF RFC 1123](https://datatracker.ietf.org/doc/html/rfc1123).
-/// Initially we'll implement a strict subset of the IETF RFC 1123, concretely:
-/// - It does not start or end with `-` or `_`.
-/// - It does not contain any characters outside of the alphanumeric range, except for `-` or `_`.
-/// - It is not empty.
-/// - It does not contain profanity.
-/// - It is not a reserved word.
+/// Initially we'll implement a strict subset of the IETF RFC 1123.
+/// Additionaly, while host segments are technically case-insensitive, the filesystem isn't,
+/// so we restrict project names to be lower case. We also restrict the use of profanity,
+/// as well as a list of reserved words.
 #[derive(Clone, Serialize, Debug, Eq, PartialEq)]
+#[cfg_attr(feature = "backend", derive(sqlx::Type, Hash))]
+#[cfg_attr(feature = "backend", sqlx(transparent))]
 pub struct ProjectName(String);
 
-impl<'de> Deserialize<'de> for ProjectName {
-    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
-    where
-        D: Deserializer<'de>,
-    {
-        let s: String = String::deserialize(deserializer)?;
-        s.parse().map_err(DeError::custom)
-    }
-}
-
-impl std::fmt::Display for ProjectName {
-    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
-        self.0.fmt(f)
+impl ProjectName {
+    pub fn new(name: &str) -> Result<Self, InvalidProjectName> {
```

Comment:
Should we rather have a method `parse` that combines both new and is_valid (returning either Self or an error)? Similarly to what is done for the api-key newtype.